Adds video, fixes structure to iOS project setup page (#10699)
flutter/website: src/content/add-to-app/ios/project-setup.md

---

Flutter UI components can be incrementally added into your existing iOS
application as embedded frameworks.
To embed Flutter in your existing application,
consider one of the following three methods.

| Embedding Method | Methodology | Benefit |
|---|---|---|
| Use CocoaPods _(Recommended)_ | Install and use the Flutter SDK and CocoaPods. Flutter compiles the `flutter_module` from source each time Xcode builds the iOS app. | Least complicated method to embed Flutter into your app. |

> | Use [iOS frameworks][] | Create iOS frameworks for Flutter components, embed them into your iOS, and update your existing app's build settings. | Doesn't require every developer to install the Flutter SDK and Cocoapods on their local machines. |

Reviewer: ```suggestion
| Use [iOS frameworks][] | Create iOS frameworks for Flutter components, embed them into your iOS, and update your existing app's build settings. | Doesn't require every developer to install the Flutter SDK and CocoaPods on their local machines. |
```

> | Use [iOS frameworks][] | Create iOS frameworks for Flutter components, embed them into your iOS, and update your existing app's build settings. Run the `flutter build` command to embed the frameworks. | Doesn't require every developer to install the Flutter SDK and Cocoapods on their local machines. |

Reviewer: Nit: `flutter build` command _creates_ the frameworks. The developer needs to embed the frameworks into the project.

Reviewer: Same comment for the row below

> | Use iOS frameworks and CocoaPods | Embed the frameworks for your iOS app and the plugins in Xcode, but distribute the Flutter engine as a CocoaPods podspec. Run the `flutter build` command to embed the frameworks. | Provides an alternative to distributing the large `Flutter.xcframework` library. |

Reviewer: Instead of saying Flutter engine, should we reuse the `Flutter.xcframework` terminology? Using both definitions of the same thing might confuse newbies.

Author: Fixed.

> {:.table .table-striped}

Reviewer: This table looks _awesome_, love it!

[iOS frameworks]: {{site.apple-dev}}/library/archive/documentation/MacOSX/Conceptual/BPFrameworks/Concepts/WhatAreFrameworks.html

> Using Flutter modules [increases your app size][].

Reviewer: We haven't introduced Flutter modules yet. Should we use simpler language here? Perhaps:

```suggestion
Adding Flutter to your existing iOS app [increases its size][].
```

For examples using an app built with UIKit,
see the iOS directories in the [add_to_app code samples][].
For an example using SwiftUI, consult the iOS directory in [News Feed App][].

## Development system requirements

Your development environment must meet the
[macOS system requirements for Flutter][] with [Xcode installed][].
Flutter supports iOS {{site.appmin.xcode}} and later and [CocoaPods][] {{site.appmin.cocoapods}} or later.

## Create a Flutter module

To embed Flutter into your existing application with any method,
create a Flutter module first.
Use the following command to create a Flutter module.

```console

> $ cd /path/to

Reviewer: This doesn't match the description below:

```suggestion
$ cd /path/to/my_flutter
```

Reviewer: @atsansone This one got missed :)

---

Flutter UI components can be incrementally added into your existing iOS
application as embedded frameworks.
To embed Flutter in your existing application,
consider one of the following three methods.

**Use the CocoaPods dependency manager and installed Flutter SDK.**
: Each time you build the app,
  it compiles the `flutter_module` from source.
  (Recommended.)

**Create frameworks for each component.**
: Create frameworks for the Flutter engine, your compiled Dart code,
  and all Flutter plugins in Xcode.
  Embed these frameworks and update your existing application's build settings.
  This method works for teams that don't want to require every developer

>   to install the Flutter SDK and Cocoapods on their local machines.

Reviewer: ```suggestion
  to install the Flutter SDK and CocoaPods on their local machines.
```

>   to install the Flutter SDK and Cocoapods on their local machines.

Reviewer: Could we move the drawbacks of option A to option A's paragraph? If you're skimming quickly, it looks like option A is recommended and has no drawbacks unless I happen to read this entire section.

**Create frameworks for Dart code and Flutter plugins. Use CocoaPods for the Flutter engine.**
: Embed the frameworks for your iOS app and the plugins in Xcode,
  but distribute the Flutter engine as a CocoaPods podspec.
  This works like the second method,

>   but it provides an alternative to distributing the large Flutter.xcframework.

Reviewer: Can we move this to option B's pragraph? See: https://github.com/flutter/website/pull/10699/files#r1626267369.

From my understanding, option B requires you to add hundreds of megabytes of binaries to source control. If that's correct, I don't think this warning is strong enough for this drawback. I _think_ most users would want option C over option B. @vashworth  @jmagman Please keep me honest here!

>   but it provides an alternative to distributing the large Flutter.xcframework.

Reviewer: Option B and C have a drawback that isn't listed here: https://github.com/flutter/flutter/issues/146957#issuecomment-2066841728

> The big limitation that Flutter has with `xcframework`s and CocoaPods is the inability to handle common dependencies, https://github.com/flutter/flutter/issues/130220. If both host app and Flutter module (Flutter module's plugin) define the same pod dependency, and Flutter module is integrated using Option B (with xcframeworks), there will be an errors like `Multiple commands produce 'CommonDependency.framework'`. The workaround is to link every plugin source (podspec) from the Flutter module directly to the host app (in Podfile), instead of using plugins' `xcframework`s produced by `build ios-framework` command, https://github.com/flutter/flutter/issues/114692. That was one of the reasons `build ios-framework --no-plugins` flag was introduced.

Should we add this `--no-plugins` flag as option D?

>   but it provides an alternative to distributing the large Flutter.xcframework.

Reviewer: Another drawback for Option C: it is incompatible with Flutter's `master` release channel. You must use the `beta` or `stable` release channels.

For examples using an app built with UIKit,
see the iOS directories in the [add_to_app code samples][].
For an example using SwiftUI, consult the iOS directory in [News Feed App][].

## Watch the video

To learn more, watch this short video on adding Flutter to an iOS app:

> <iframe class="full-width" src="{{site.yt.embed}}/IIcrfrTshTs" title="Learn about how to add Flutter to an existing iOS app" {{site.yt.set}}></iframe>

Reviewer: This video covers method A only. Should we move it to method A's section?

## System requirements

Your development environment must meet the
[macOS system requirements for Flutter][] with [Xcode installed][].

> Flutter supports iOS {{site.appmin.xcode}} and later and [CocoaPods][] {{site.appmin.cocoapods}} or later.

Reviewer 2: ```suggestion
Flutter supports iOS {{site.appmin.ios}} and later and [CocoaPods][] {{site.appmin.cocoapods}} or later.
```

But why is this 17?   Flutter supports iOS 12.
https://github.com/flutter/website/blob/b6af70700133b17968a7a4484455b300f9111c6b/src/_data/site.yml#L77

Reviewer 2: Oh, I see this should be `site.targetmin.ios`?

Reviewer 2: This one wasn't fixed.
<img width="538" alt="Screenshot 2024-06-13 at 4 56 01 PM" src="https://github.com/flutter/website/assets/682784/b33c742f-2afb-499f-a5b0-24bf9f1eb047">

Author: Bigger issue was that this should have said Xcode 14 or later, not iOS. 🫨

## Create a Flutter module

To embed Flutter into your existing application with any method,
create a Flutter module first.
Use the following command to create a Flutter module.

```console
$ cd /path/to
$ flutter create --template module my_flutter
```

Flutter creates module project under `/path/to/my_flutter/`.

> If you use Method A, the module should be created

Reviewer: FYI, at this point in the guide `Method A` hasn't been defined yet.

in the same parent directory as your existing iOS app.

From the Flutter module directory, you can run the same `flutter`
commands you would in any other Flutter project,

> like `flutter run --debug` or `flutter build ios`.

Reviewer: Should we drop the `--debug` option? It's not clear to me what value it adds

```suggestion
like `flutter run` or `flutter build ios`.
```

Reviewer: Re-opening this as the current text still has `--debug`. Is there a reason to keep the `--debug` option here? I don't feel strongly but I do think it's a little bit better :)

      **Info-Release.plist** in **Release**.

      {% render docs/captioned-image.liquid,
      image:"development/add-to-app/ios/project-setup/plist-build-setting.png",
      caption:"The updated **Info.plist File** build setting displaying the
      configuration variations" %}

1. Remove the **Release** property list from the **Build Phases**.

   1. In the **Navigator**, click on your project.

   1. Click the **Build Phases** tab.

   1. Expand **Copy Bundle Resources**.

   1. If this list includes `Info-Release.plist`,
      click on it and then click the **-** (minus sign) under it
      to remove the property list from the resources list.

      {% render docs/captioned-image.liquid,
      image:"development/add-to-app/ios/project-setup/copy-bundle.png",
      caption:"The **Copy Bundle** build phase displaying the
      **Info-Release.plist** setting. Remove this setting." %}

1. The first Flutter screen your Debug app loads prompts
   for local network permission.

   Click **OK**.

   _(Optional)_ To grant permission before the app loads, enable
   **Settings > Privacy > Local Network > Your App**.

## Mitigate known issue with Apple Silicon Macs

> On an Apple Silicon (M1 or M2) Mac,

Reviewer: ```suggestion
On an [Apple Silicon macs](https://support.apple.com/en-us/116943),
```

There's M3 macs now! There will likely be M4 soon too, as that was recently released for iPad. Perhaps we should just link to Apple's list of Apple Silicon devices instead?

    </li>
</ul>

{% comment %} Tab panes {% endcomment -%}
<div class="tab-content">

<div class="tab-pane active" id="embed-using-cocoapods" role="tabpanel" aria-labelledby="embed-using-cocoapods-tab">

{% render docs/add-to-app/ios-project/embed-cocoapods.md, yt-embed:
site.yt.embed, yt-set: site.yt.set, yt-watch: site.yt.watch  %}

</div>

<div class="tab-pane" id="embed-using-frameworks" role="tabpanel" aria-labelledby="embed-using-frameworks-tab">

{% render docs/add-to-app/ios-project/embed-frameworks.md %}

</div>
<div class="tab-pane" id="embed-using-both" role="tabpanel" aria-labelledby="embed-using-both-tab">

{% render docs/add-to-app/ios-project/embed-split.md %}

</div>
</div>
{% comment %} End: Tab panes. {% endcomment %}

## Set local network privacy permissions

On iOS 14 and later, enable the Dart multicast DNS service in the
**Debug** version of your iOS app.
This adds [debugging functionalities such as hot-reload and DevTools][]
using `flutter attach`.

:::warning
Never enable this service in the **Release** version of your app.
The Apple App Store might reject your app.
:::

> Maintain a separate copy of your app's Info.plist per build configuration.

Reviewer: ```suggestion
Maintain a separate copy of your app's `Info.plist` per build configuration.
```

directory don't appear in your existing iOS project
using the module, and might be overwritten by Flutter.

Exclude the `.ios/` directory from source control as it's autogenerated.

> Before building the module on a new machine, run `flutter pub get`

Reviewer: This is really important information, but currently it's buried as the second sentence in the second paragraph of a note. Should we make this it's own separate note?

in the `my_flutter` directory first.

> This regenerates the `.ios/` directory before building iOS project

Reviewer: ```suggestion
This regenerates the `.ios/` directory before building the iOS project
```

using the Flutter module.
:::

## Embed a Flutter module in your iOS app

After you have developed your Flutter module,
you can embed it using the methods described at the top of the page.

You can run in Debug mode on a simulator or a real device,
and Release on a real device.

> Using Flutter [increases your app size][].

Reviewer: This warning seems a bit late. Should we move this up to the intro?

:::note
Learn more about [Flutter's build modes][build modes of Flutter].

To use Flutter debugging features such as hot reload,
consult [Debugging your add-to-app module][].
:::

> ### Method A - Embed with CocoaPods and the Flutter SDK {:#method-a}

Reviewer 2: I've always found this "Option A/Option B/Option C" list awkward and makes the page very long. Is there any existing website doc semantics for a decision tree? Or make this tabbed so you only see one at a time?  Could use tech writer expertise for advice on how to communicate this better!

Reviewer: A huge +1 to this. I would love to see tabs like on the installation guide: https://docs.flutter.dev/get-started/install/macos/mobile-ios?tab=ios16#install-the-flutter-sdk

Author: No issue with doing that. It does cause some issues with inline linking, but nothing too offputting.

#### Requirements {:#method-a-reqs}

Every developer working on your project must have a local version
of the Flutter SDK and CocoaPods installed.

#### Approach {:#method-a-approach}

> This first method compiles the Flutter module from source each time
> Xcode builds the app. To run the script to embed your Dart and plugin code,
> build your iOS app in Xcode.

Reviewer: Should we drop the second sentence? It's original purpose was to reveal how the sausage was made. Perhaps it would be better to simplify this further?

```suggestion
This first method compiles your Dart and plugin code from source each time
Xcode builds the app.
```

    └── Podfile
```

#### Update your Podfile

1. If your existing application (`MyApp`) doesn't already have a Podfile,
   use the `pod init` command in the `MyApp` directory to create one.

   To learn more about CocoaPods,
   consult the [CocoaPods getting started guide][].

> 1. Add the following lines to your `Podfile`.

Reviewer: Can we clarify where line should be added?

This was unclear to me when I followed this guide. I ended up looking at our samples on GitHub to figure this out. See: https://docs.google.com/document/d/1fnmxAHTLhYTmssi8YI55av7PRXjFPNutOrEAxGrYIEI/edit?resourcekey=0-A9xndfJnQCTdRL0QtOqRig&tab=t.0#heading=h.uxxukmyfol1w

```suggestion
1. Add the following lines at the top of your `Podfile`.
```

   ```ruby title="MyApp/Podfile"
   flutter_application_path = '../my_flutter'
   load File.join(flutter_application_path, '.ios', 'Flutter', 'podhelper.rb')
   ```

1. For each [Podfile target][] that needs to embed Flutter,
   add a call to `install_all_flutter_pods(flutter_application_path)`.

   ```ruby title="MyApp/Podfile"
   target 'MyApp' do
     install_all_flutter_pods(flutter_application_path)
   end
   ```

1. In the `Podfile`'s `post_install` block,
   add a call to `flutter_post_install(installer)`.

   ```ruby title="MyApp/Podfile"
   post_install do |installer|
     flutter_post_install(installer) if defined?(flutter_post_install)
   end
   ```

>    The `flutter_post_install` method adds build settings to support

Reviewer: A new `Podfile` does not have a `post_install` block. Should we add a note on this scenario? Perhaps:

```suggestion
   :::note
   If your `Podfile` does not have a `post_install` block yet, add this code to the bottom of your `Podfile`.
   :::
```

   native Apple Silicon `arm64` iOS simulators.
   Include the `if defined?(flutter_post_install)`
   check to validate your `Podfile` if you are running on older versions

>    of Flutter that lack this method.

Reviewer: Should we remove this explanation? It's not clear to me why the user needs to know this or what they might do differently with this information.

>    of Flutter that lack this method.

Reviewer: When I followed these steps, I was confused _where_ I should add content. I would've have liked a reference `Podfile` to mirror.

Should we link to a reference `Podfile` here? Or perhaps should we embed an example `Podfile` before and after these changes?

1. Use the `flutter pub get` command to refresh plugins.

   When you change the Flutter dependencies in the `pubspec.yaml` file,
   run `flutter pub get` in your Flutter module directory.
   This refreshes the list of plugins that the `podhelper.rb` script reads.

1. Run the `pod install` command to embed the plugins and frameworks.

   From your iOS app project at `/path/to/MyApp`, run `pod install`.

   The `podhelper.rb` script embeds your plugins,

>    `Flutter.framework`, and `App.framework` into your project.

Reviewer: Should we move this up to the step that adds the `podhelper.rb` script to your `Podfile`?

   Your app's **Debug** and **Release** build configurations embed
   the **Debug** or **Release** [build modes of Flutter][], respectively.

   To test in profile mode, add a **Profile** build configuration to your app.

   :::tip
   `Flutter.framework` is the bundle for the Flutter engine,

>    and `App.framework` is the compiled Dart code for this project.

Reviewer: Consider explaining where plugins go too:

```suggestion
   and `App.framework` is the compiled Dart and plugin code for this project.
```

I don't feel strongly, feel free to skip for simplicity

   :::

> 1. Open `MyApp.xcworkspace` in Xcode.

Reviewer: A very common mistake is to open `MyApp.xcodeproj` instead of `MyApp.xcworkspace`: the `.xcodeproj` won't build as it doesn't have CocoaPod dependencies. Here's a recent user interview that specifically called this out as a top pain point: [Google internal](https://docs.google.com/document/d/1X9dNGdhBIxKXnHxGE1UqRZ0TOwmsNZw4svdu3JWZP_o/edit?usp=sharing&resourcekey=0-HTwX-TDNA80Uf7-TLjTYKw).

Your app won't have an `.xcworkspace` if it wasn't using CocoaPods before; the `.xcworkspace` is created by `pod install`. These developers will need to switch their workflow to open the `.xcworkspace` instead of the `.xcodeproj`.

Should we add more details to this section to call this out?

```

:::warning
Always put `Flutter.xcframework` and `App.xcframework` from the same directory.
Mixing `.xcframework` imports from different directories
(like `Profile/Flutter.xcframework` with `Debug/App.xcframework`)
causes runtime crashes.
:::

#### Procedures

You can either link or embed the generated frameworks
into your existing app in Xcode.
Use the method that is best for your project.

##### Link on the frameworks

For example, you can drag the frameworks from
`/path/to/MyApp/Flutter/Release/` in the Finder into your target's

> **Build Settings > Build Phases > Link Binary With Libraries**.

Reviewer: Can we add a screenshot for this step? For example:

![image](https://github.com/flutter/website/assets/737941/c2977d68-484b-4a30-baee-056093f80765)

to the **Framework Search Paths** (`FRAMEWORK_SEARCH_PATHS`).

{% render docs/captioned-image.liquid, image:"development/add-to-app/ios/project-setup/framework-search-paths.png", caption:"Update Framework Search Paths in Xcode" %}

:::tip
To use the simulator, embed the Debug version of the Flutter frameworks
in your Debug build configuration.
To do this, use `$(PROJECT_DIR)/Flutter/$(CONFIGURATION)`
in the **Framework Search Paths** (`FRAMEWORK_SEARCH_PATHS`) build setting.
This embeds the Release frameworks in the Release configuration,
and the Debug frameworks in the Debug Configuration.

You must also open `MyApp.xcodeproj/project.pbxproj` (from Finder)
and replace `path = Flutter/Release/example.xcframework;`
with `path = "Flutter/$(CONFIGURATION)/example.xcframework";`
for all added frameworks. The double quotes (`"`) are necessary.
:::

##### Embed the frameworks

To load the generated dynamic frameworks at runtime,
embed them into your app.

:::important
Plugins might produce [static or dynamic frameworks][].
Static frameworks should be linked on, but never embedded.
If you embed a static framework into your iOS app,
you can't publish that app to the App Store.

> Publishing fails with a

Reviewer: Should this be in future simple tense?

```suggestion
Publishing will fail with a
```

Author: @loic-sharma : Google Style is present whenever possible. With a slight change, it is.

**Found an unexpected Mach-O header code** archive error.
:::

> After linking the frameworks, they should display in the

Reviewer: Consider "appear" to make it clearer that this won't happen until after you add linking:

```suggestion
After linking the frameworks, they should appear in the
```

**Frameworks, Libraries, and Embedded Content**
section of your target's **General** settings.

To embed the dynamic frameworks, select **Embed & Sign**.
They then appear under **Embed Frameworks** within **Build Phases** as shown.

{% render docs/captioned-image.liquid, image:"development/add-to-app/ios/project-setup/embed-xcode.png", caption:"Embed frameworks in Xcode" %}

You should now be able to build the project in Xcode using `⌘B`.

> ### Method C - Embed app and plugin frameworks in Xcode; Flutter framework with CocoaPods {:#method-c}

Reviewer 2: I'm not sure why but this caused some wrapping and `#` cropping on macOS Chrome:
![Screenshot 2024-06-03 at 4 24 32 PM](https://github.com/flutter/website/assets/682784/8209dcc9-6659-46e2-a5ee-b52c6df33cb2)

#### Requirements {:#method-c-reqs}

> This method has the same requirements as Method B.

Reviewer: Could we inline the requirements to this section? The sidebar doesn't have a link to Method B's requirements.

#### Approach {:#method-c-approach}

> This method generates Flutter as a CocoaPods podspec instead of

Reviewer: Should this have backticks?

```suggestion
This method generates Flutter as a CocoaPods `.podspec` instead of
```

I don't feel strongly

## Local Network Privacy Permissions

On iOS 14 and later, enable the Dart multicast DNS service in the
Debug version of your app to add
[debugging functionalities such as hot-reload and DevTools][]
using `flutter attach`.

:::warning
This service must not be enabled in the **Release**
version of your app, or you might experience App Store rejections.
:::

Maintain a separate copy of your app's Info.plist per build configuration.
The following instructions assume the default **Debug** and **Release**.
Adjust the names as needed depending on your app's build configurations.

> 1. Rename your app's **Info.plist** to **Info-Debug.plist**.

Reviewer: These instructions are hard to follow if you have a new Swift app as they don't have a `Info.plist` file and Xcode's add file dialog doesn't have a `.plist` option.

Consider adding a note like:

```
SwiftUI projects start without an `Info.plist` file. You can create one by going to your target's `Info` > `Custom iOS Target Properties` and editing a property.
```

1. Copy the plist and call it **Info-Release.plist**.

1. Add **Info-Release.plist** to your Xcode project.

    {% render docs/captioned-image.liquid,
    image:"development/add-to-app/ios/project-setup/info-plists.png",
    caption:"Info-Debug.plist and Info-Release.plist in Xcode" %}

1. In **Info-Debug.plist** _only_ add the key `NSBonjourServices`.

> 1. Set the value to an array with the string `_dartVmService._tcp`.

Reviewer: Can we make `_dartVmService._tcp` a separate code block so that there's a copy button? Selecting this text to copy it is a little annoying.
